File: Maximum-Product-of-Word-Lengths/MPoWL.rs
//use std::collections::HashSet;

pub fn max_product(words: Vec<String>) -> i32 {
    use std::collections::HashSet;
    let words_set = words
        .iter()
        .map(|w| w.chars().collect::<HashSet<_>>())
        .collect::<Vec<_>>();

    let mut most = 0;
    for i in 0..words_set.len() {
        for j in i + 1..words_set.len() {
            if words_set[i].intersection(&words_set[j]).count() == 0 {
                if words[i].len() * words[j].len() > most {
                    most = words[i].len() * words[j].len();
                }
            }
        }
    }

    most as i32
}

fn main() {
    dbg!(max_product(vec![
        "abcw".to_string(),
        "baz".to_string(),
        "foo".to_string(),
        "bar".to_string(),
        "xtfn".to_string(),
        "abcdef".to_string()
    ]));

    dbg!(max_product(vec![
        "a".to_string(),
        "ab".to_string(),
        "abc".to_string(),
        "d".to_string(),
        "cd".to_string(),
        "bcd".to_string(),
        "abcd".to_string(),
    ]));
}

File: Maximum-Product-of-Word-Lengths/MPoWL.ml
(* whooo *)
let string_to_chars s : char list = s |> String.to_seq |> List.of_seq
let string_to_chars_seq s = s |> String.to_seq

module CharSet = Set.Make (Char)

(* make sets of words *)
let rec make_char_set words =
  match words with
  | word :: ws ->
      let cs = CharSet.of_list (string_to_chars word) in
      cs :: make_char_set ws
  | [] -> []

let max_product words =
  let words_sets = make_char_set words in
  let most = ref 0 in
  for i = 0 to List.length words_sets - 1 do
    for j = i + 1 to List.length words_sets - 1 do
      if
        CharSet.is_empty
          (CharSet.inter (List.nth words_sets i) (List.nth words_sets j))
      then
        most :=
          max !most
            ((List.nth words_sets i |> CharSet.to_list |> List.length)
            * (List.nth words_sets j |> CharSet.to_list |> List.length))
    done
  done;
  !most
;;

max_product [ "abcw"; "baz"; "foo"; "bar"; "xtfn"; "abcdef" ] = 16;;
max_product [ "a"; "ab"; "abc"; "d"; "cd"; "bcd"; "abcd" ] = 4
